haystack/dataclasses/chat_message.py

```diff
@@ -5,104 +5,306 @@
 import warnings
 from dataclasses import asdict, dataclass, field
 from enum import Enum
-from typing import Any, Dict, Optional
+from typing import Any, Dict, List, Optional, Sequence, Union
+
+LEGACY_INIT_PARAMETERS = {"role", "content", "meta", "name"}
 
 
 class ChatRole(str, Enum):
-    """Enumeration representing the roles within a chat."""
+    """
+    Enumeration representing the roles within a chat.
+    """
 
-    ASSISTANT = "assistant"
+    #: The user role. A message from the user contains only text.
     USER = "user"
+
+    #: The system role. A message from the system contains only text.
     SYSTEM = "system"
+
+    #: The assistant role. A message from the assistant can contain text and Tool calls. It can also store metadata.
+    ASSISTANT = "assistant"
+
+    #: The tool role. A message from a tool contains the result of a Tool invocation.
+    TOOL = "tool"
+
+    #: The function role. Deprecated in favor of `TOOL`.
     FUNCTION = "function"
 
+    @staticmethod
+    def from_str(string: str) -> "ChatRole":
+        """
+        Convert a string to a ChatRole enum.
+        """
+        enum_map = {e.value: e for e in ChatRole}
+        role = enum_map.get(string)
+        if role is None:
+            msg = f"Unknown chat role '{string}'. Supported roles are: {list(enum_map.keys())}"
+            raise ValueError(msg)
+        return role
+
+
+@dataclass
+class ToolCall:
+    """
+    Represents a Tool call prepared by the model, usually contained in an assistant message.
+
+    :param id: The ID of the Tool call.
+    :param tool_name: The name of the Tool to call.
+    :param arguments: The arguments to call the Tool with.
+    """
+
+    tool_name: str
+    arguments: Dict[str, Any]
+    id: Optional[str] = None  # noqa: A003
+
+
+@dataclass
+class ToolCallResult:
+    """
+    Represents the result of a Tool invocation.
+
+    :param result: The result of the Tool invocation.
+    :param origin: The Tool call that produced this result.
+    :param error: Whether the Tool invocation resulted in an error.
+    """
+
+    result: str
```

Comment:
A note here - I think we should have the result field be a Union of TextContent and other future media types that will include ByteStream and perhaps the upcoming MediaContent. Not sure if it is smart to add that immediately i.e have:
```
result: TextContent
```
and then make it Union as we add those new types. Or add it all in one batch sometimes soon when introduce these new types. wdyt is the best approach here?

Reply:
I haven't studied in depth the multimodal support in different API providers, so I would prefer to introduce these changes when we know more, hopefully in a non-breaking way.

Reply:
Yeah, neither did I  but @mathislucka mentioned encountering this hurdle in his experiments IIRC. If it is a big change to have TextContent here then perhaps lets leave it for next release as this is admittedly already a big change. Wonder what do you think about this @LastRemote - i.e. having tools returning non str content. Have you already encountered that use case?

Reply:
> I haven't studied in depth the multimodal support in different API providers, so I would prefer to introduce these changes when we know more, hopefully in a non-breaking way.

@anakin87 I did some research on this previously: Anthropic models can accept image (as an object containing base64 and mime type) tool result as well as text.

Reply:
> having tools returning non str content. Have you already encountered that use case?

@vblagoje I am aware of this but it is not my current use case. To be honest I haven't really thought about this (I didn't change it in my multimodal PR), but this could be another `Sequence[ChatMessageContentT]` for maximum flexibility (it is up to the models and corresponding utility functions whether to accept certain types in tool call result)

Reply:
Yes, something like that. As this future change will likely break specific tool integrations, and not affect many users, perhaps we can postpone it to after the integration and keep str as ToolCallResult field for now

Reply:
> ...and not affect many users...

I prefer being a little careful on this one though. If a future change makes `result` to be a list (Anthropic already supports this), existing tool implementations may be affected.

Reply:
In all honesty, I would prefer to keep the PR as it is regarding this aspect, especially since these changes have been extensively tested in recent months.

I am afraid that introducing a change on the fly at this time could cause some bugs, as well as complicate our lives.

For the future, unlike this PR, I think we can introduce the changes gradually (and without breaking changes), because they would be much smaller in scope.

(@vblagoje regarding the problems Mathis encountered, my impression was that they were not directly related to the tool message types, but to the `Toolnvoker` return types.)

Reply:
> In all honesty, I would prefer to keep the PR as it is regarding this aspect...

Sure, I am fine with it. Right now I am not seeing a lot of valid use cases for this, and there is still room for future updates.

Reply:
Yes @anakin87 agreed